trl/trainer/gkd_trainer.py

```diff
@@ -0,0 +1,211 @@
+        kl_student = F.kl_div(interpolated_log_probs, student_log_probs, reduction="none", log_target=True)
+
+        # Compute the Generalized Jensen-Shannon Divergence
+        jsd = beta * kl_student + (1 - beta) * kl_teacher
+
+        # Apply reduction
+        if reduction == "batchmean":
+            return jsd.sum() / (jsd.size(0) * jsd.size(1))
+        elif reduction == "sum":
+            return jsd.sum()
+        elif reduction == "mean":
+            return jsd.mean()
+        else:
+            return jsd
+
+    def compute_loss(self, model, inputs, return_outputs=False):
+        # compute student output
+        outputs_student = model(
+            input_ids=inputs["input_ids"], attention_mask=inputs["attention_mask"], labels=inputs["labels"]
+        )
+
+        # compute teacher output in eval mode
+        self.teacher_model.eval()
+        with torch.no_grad():
+            outputs_teacher = self.teacher_model(
+                input_ids=inputs["input_ids"], attention_mask=inputs["attention_mask"], labels=inputs["labels"]
+            )
+
+        # compute loss
+        loss = 0
+        batch_size = inputs["labels"].shape[0]
+        for i in range(batch_size):
+            index = (inputs["labels"] != -100)[i]
+            student_logits = outputs_student.logits[i, index, :][-self.args.max_new_tokens_response :, :]
+            teacher_logits = outputs_teacher.logits[i, index, :][-self.args.max_new_tokens_response :, :]
+            loss += self.generalized_jsd_loss(
+                student_logits=student_logits,
+                teacher_logits=teacher_logits,
+                beta=self.beta,
+            )
+        loss /= batch_size
+
+        # Return loss
+        return (loss, outputs_student) if return_outputs else loss
+
+    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]) -> torch.Tensor:
+        """
+        Perform a training step on a batch of inputs.
+
+        Subclass and override to inject custom behavior.
+
+        Args:
+            model (`nn.Module`):
+                The model to train.
+            inputs (`Dict[str, Union[torch.Tensor, Any]]`):
+                The inputs and targets of the model.
+
+                The dictionary will be unpacked before being fed to the model. Most models expect the targets under the
+                argument `labels`. Check your model's documentation for all accepted arguments.
+
+        Return:
+            `torch.Tensor`: The tensor with training loss on this batch.
+        """
+        if random.random() >= self.lmbda:
+            # On-policy: Generate outputs from the student model with temperature self.temperature
+            with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
+                generation_config = GenerationConfig(
+                    max_new_tokens=self.args.max_new_tokens_response,
+                    temperature=self.temperature,
+                    do_sample=True,
```

Comment:
If I'm not mistaken, we need to override the default value for top_k to do pure sampling: https://huggingface.co/docs/transformers/v4.44.0/en/main_classes/text_generation#transformers.GenerationConfig

```suggestion
                    do_sample=True,
                    top_k=0,
```